lib/surface/components/form/error_tag.ex

```diff
@@ -0,0 +1,124 @@
+  which utilizes `Gettext`. You need to provide a tuple with the module and
+  the name of the function (as an atom) as follows:
+
+  ```elixir
+  config :surface, :components, [
+    {Surface.Components.Form.ErrorTag, translator: {MyAppWeb.ErrorHelpers, :translate_error}}
+  ]
+  ```
+
+  There is also a `translator` prop which can be used on a case-by-case basis.
+  It overrides the configuration.
+
+  ## Examples
+
+  ```surface
+  <ErrorTag field="password" />
+  ```
+
+  ```surface
+  <Field name="password>
+    <ErrorTag />
+  </Field>
+  ```
+
+  ```surface
+  <ErrorTag phx_feedback_for="confirm_password_for_reset" />
+  ```
+
+  ```surface
+  <ErrorTag class="custom-css-classes" />
+  ```
+
+  ```surface
+  <ErrorTag translator={{ &CustomTranslationLib.translate_error/1 }} />
+  ```
+  """
+
+  use Surface.Component
+
+  import Phoenix.HTML.Form, only: [input_id: 2]
+
+  alias Surface.Components.Form.Input.InputContext
+
+  @doc "An identifier for the form"
+  prop form, :form
+
+  @doc "An identifier for the associated field"
+  prop field, :atom
+
+  @doc "Class or classes to apply to each error tag <span>"
+  prop class, :css_class
+
+  @doc """
+  A function that takes one argument `{msg, opts}` and returns
+  the translated error message as a string. If not provided, falls
+  back to Phoenix's default implementation.
+
+  This can also be set via config, for example:
+
+  ```elixir
+  config :surface, :components, [
+    {Surface.Components.Form.ErrorTag, translator: {MyApp.Gettext, :translate_error}}
+  ]
+  ```
+  """
+  prop translator, :fun
+
+  @doc """
+  If you changed the default ID on the input, provide it here.
+  (Useful when there are multiple forms on the same page, each
+  with an input of the same name. LiveView will exhibit buggy behavior
+  without assigning separate id's to each.)
+  """
+  prop phx_feedback_for, :string
+
+  def render(assigns) do
+    translate_error = assigns.translator || translator_from_config() || (&translate_error/1)
+
+    ~H"""
+    <InputContext assigns={{ assigns }} :let={{ form: form, field: field }}>
+      <span
+        :for={{ error <- Keyword.get_values(form.errors, field) }}
+        class={{ @class }}
+        phx-feedback-for={{ @phx_feedback_for || input_id(form, field) }}
+      >{{ translate_error.(error) }}</span>
+    </InputContext>
+    """
+  end
+
+  @doc """
+  Translates an error message.
+
+  This is the fallback (Phoenix's default implementation) if a translator
+  is not provided via config or the `translate` prop.
+  """
+  def translate_error({msg, opts}) do
+    # Because the error messages we show in our forms and APIs
+    # are defined inside Ecto, we need to translate them dynamically.
+    Enum.reduce(opts, msg, fn {key, value}, acc ->
+      String.replace(acc, "%{#{key}}", to_string(value))
+    end)
+  end
+
+  defp translator_from_config do
+    if translator = get_config(:translator) do
```

Comment:
One thing I've just noticed that I've been using the `default_` prefix on those kinds of configuration. For instance:

```elixir

config :surface, :components, [
  {Surface.Components.Markdown, default_class: "content markdown", default_opts: [smartypants: false]},
  # Bulma defaults
  {Surface.Components.Form.Input, default_class: "input"},
  {Surface.Components.Form.Field, default_class: "field"},
  {Surface.Components.Form.TextArea, default_class: "textarea"},
  {Surface.Components.Form.Label, default_class: "label"}
]
```

The intention was to be more explicit that those values are just defaults since the component author can do whatever he wants with any of those values, e.g. use it as a default, override an existing value, merge the value (like Markdown's opts), etc. 

However, I think this might be a silly concern and I we should probably remove the prefixes and keep everything with the same name. Thoughts?

Reply:
> since the component author can do whatever he wants with any of those values

I think that's the rub. If the component uses it as a default that can be overridden with a prop, personally I think `default_` potentially adds clarity.

But if it merges the config value with prop value, I'm not sure `default_` is the most clear. I can't think of anything at the moment more clear in that instance than removing the prefix and documenting the merge functionality explicitly.

That said, I think `default_translator` would be an improvement here. Shall we move forward with that?

Related question: where is the standard place to document these configuration options? I put it in the `moduledoc` as well as the related `prop` because I couldn't find another relevant place.

Reply:
I agree re default adding clarity, though I'm not quiet following the distinction between merging and overriding - would that only apply for a keyword list like opts?

Reply:
> I think that's the rub. If the component uses it as a default that can be overridden with a prop, personally I think `default_` potentially adds clarity.
> 
> But if it merges the config value with prop value, I'm not sure `default_` is the most clear. I can't think of anything at the moment more clear in that instance than removing the prefix and documenting the merge functionality explicitly.

That's exactly why I changed my mind. I think we shouldn't try to have standard prefixes that are semantically consistent across all components. Since each component can do whatever it wants, it should be up to the author to properly document how the config is used.

> where is the standard place to document these configuration options? I put it in the `moduledoc` as well as the related `prop` because I couldn't find another relevant place.

Yeah, `moduledoc` is the right place for now. I thought about adding something like `config` to the API. Like:

```elixir
@doc "Class or classes to apply to each error tag <span>"
prop class, :css_class

@doc "The default class"
config default_class, :string
```

That would make configs first-class citizens but I'm not sure we should do this. Configs should not be abused as it makes the component appearance and behaviour much less intuitive as you change projects. By promoting them to the API, I fear people will start using configs everywhere.

Reply:
@msaraiva 100% agree about trying not to encourage usage of config. I see what you're saying.

@lnr0626 I agree, I think it'd typically only be mergeable with composite values like keyword lists.

Reply:
Ok, I changed the config key from `translator` to `default_translator`.

Reply:
Fair enough. Let's keep the `default_` prefix then.